keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,728 @@
+- Implement the feature as an advisory feature
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+- Implement any of the above potential use cases
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user knows this is an important object, and that terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens`.
+
+`Liens` is defined as a slice of strings, like `Finalizers`.
+The strings need to be namespaced keys.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
+`Liens` should be added per controller or per user basis.
+Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
+(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).
+
+A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
+Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).
+
+### API Changes
+
+`Liens` is added to `ObjectMeta`.
+
+```go
+Liens []string
+```
+
+Validation criteria of the field are as follows:
+- Keys must be namespaced like "kubernetes.io/secret-protection" and "foo/bar".
+- Maximum length of each key is 64 characters
```

Comment:
64 may be tight - I think we allow 253 for most label-like strings

Reply:
+1